### DISCUSSION TITLE
Error! in Parent Austenite Grains Reconstruction from Martensite using MTEX

### DISCUSSION THREAD

**Discussion by AmitS07** [mtex-toolbox/mtex#586] [Ask Anything] [answered]
Hello All,

I'm facing a problem while reconstructing the  Parent Grains Reconstruction from Martensite through MTEX. The errors are as per the attached figure below. 

![image](https://user-images.githubusercontent.com/67289970/99930961-dfdc1500-2d78-11eb-9005-06088470a04a.png)
![image](https://user-images.githubusercontent.com/67289970/99930971-e79bb980-2d78-11eb-8048-2596b8bca27d.png)

Hence, may I know how to solve the above error?

**Comment by nyyssont** [comment, marked as answer]
Hello, the above error would indicate you are attempting to do something with a dataset that contains information for two phases but is only allowed for one phase.

The below error is due to the fact that you have ebsd.grainId set as output but ebsd('Iron (Alpha)','indexed') set as input. ebsd.grainId contains both indexed and unindexed points, while the latter contains only indexed points.

As a first remedy for the script below, I would try changing the calcGrains commands to the following form:

    [grains, ebsd('indexed').grainId] = calcGrains(ebsd('indexed'),'angle',3*degree)

if you wish to have single-phase grain map, you need to have:

    [grains, ebsd('Iron bcc').grainId] = calcGrains(ebsd('Iron bcc'),'angle',3*degree)

It might help if you disable your other phases from the map (you may restore them later) before constructing grains. Example for austenite:

    ebsd('Iron fcc').rotations = quaternion.nan;
    ebsd('Iron fcc').phase = 0;

**Reply by AmitS07** [reply]
Dear Tuomo,

Thank you very much for your answer.

**Comment by AmitS07** [comment]
Dear Dr. Tuomo,

In the continuation of the Parent Grains Construction from Martensite, now I'm having another error and It is coming especially in Computing Child Variants. The script line is as per follows-
% consider only martensite pixels that now belong to austenite grains
**isNowFCC = parentGrains.phaseId(max(1,parentEBSD.grainId)) == 3 & parentEBSD.phaseId == 2;**

and the error is

**Index exceeds array bounds**

Error in grain2d/subsref (line 44)
[varargout{1:nargout}] = builtin('subsref',grains,s);

How to address the above error? 

Thanks in advance

**Comment by AmitS07** [comment]
Dear Dr. Tuomo and Prof. Ralf,

It is a continuation to the PAG_reconstructions. Now, with the help of **Martensite Parent Grain Reconstruction** script and above mentioned script line `[grains, ebsd('Iron bcc').grainId] = calcGrains(ebsd('Iron bcc'),'angle',3*degree)`  I'm able to reconstruct the Austenite Phase, which can be seen from below images. 

Fig.1. Martensite (Iron (Alpha) IPF image
![image](https://user-images.githubusercontent.com/67289970/100341480-9b63aa00-3002-11eb-9720-cb44ec13cd10.png)

Fig.2. Austenite (Iron (Gamma) IPF image after PAG reconstruction
![image](https://user-images.githubusercontent.com/67289970/100340331-0318f580-3001-11eb-8ce1-e918becc3e9e.png)

It is clear from Fig.2 that reconstruction is working properly. Whereas it can be seen that lot of noise is still present. I "denoise" the data to reduce it. But it seems further smoothening will be require. hence, my question is that how to do that?

Thank you very much for this beautiful software "MTEX".

Sincerely
Amit

**Comment by ralfHielscher** [comment]
You may want to check the command
```matlab
grains = merge(grains,'inclusions')
```
which merges all inclusions into the host grains.

Ralf.

**Reply by AmitS07** [reply]
Dear Prof. Ralf,

Thank you very much for your comment on inclusion i.e. smoothing the parent data.

Yours sincerely

Amit

**Comment by AmitS07** [comment]
Dear Dr. Tuomo and Prof. Ralf,

I am again getting an error as mentioned earlier for the below line.
`isNowFCC = parentGrains.phaseId(max(1,parentEBSD.grainId)) == 3 & parentEBSD.phaseId == 2;`
and the Error is
`Index exceeds array bounds`
`Error in grain2d/subsref (line 44)`
` [varargout{1:nargout}] = builtin('subsref',grains,s);`

I only changed the EBSD scan data and I follow the follows lines to make it single phase-
  `[grains, ebsd('Iron(Alpha)').grainId] = calcGrains(ebsd('Iron(Alpha)'),'angle',3*degree);`

Actually, the microstructure contains 99.6% Iron (Alpha) phase. Hence, I'm only considering Iron(Alpha) phase.

If possible, kindly address the above error. I'm not able solve it.

Thanks in advance

------------------
Amit
